File: src/plnflr/engine/grid.py
"""Axis-aligned rectangular grid on the inset bounding box."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal


@dataclass(frozen=True, slots=True)
class Slot:
    row_index: int
    x0: int
    y0: int
    x1: int
    y1: int
    kind: Literal["full", "start_cut", "end_cut", "rip"]


def split_rows(inner_mm: int, plank_width_mm: int, min_row_width_mm: int) -> tuple[int, ...]:
    if inner_mm <= 0 or plank_width_mm <= 0:
        raise ValueError("spans must be positive")
    n, rem = divmod(inner_mm, plank_width_mm)
    if n == 0:
        return (inner_mm,)
    if rem == 0:
        return (plank_width_mm,) * n
    if rem < min_row_width_mm:
        extra = plank_width_mm + rem
        first = extra // 2
        last = extra - first
        middle = n - 1
        if middle < 0:
            return (first, last) if last else (first,)
        return (first,) + (plank_width_mm,) * middle + (last,)
    return (plank_width_mm,) * n + (rem,)


def _row_lengths(
    inner_mm: int,
    plank_length_mm: int,
    start_cut_mm: int,
    min_end_mm: int,
) -> tuple[int, ...]:
    start = start_cut_mm
    if start <= 0 or start >= plank_length_mm:
        start = plank_length_mm
    remaining = inner_mm - start
    if remaining < 0:
        return (inner_mm,)
    lengths = [start]
    while remaining > 0:
        if remaining >= plank_length_mm:
            lengths.append(plank_length_mm)
            remaining -= plank_length_mm
        else:
            lengths.append(remaining)
            remaining = 0
    if len(lengths) >= 2 and lengths[-1] < min_end_mm:
        need = min_end_mm - lengths[-1]
        if lengths[0] - need >= min_end_mm:
            lengths[0] -= need
            lengths[-1] += need
        else:
            extra = lengths[0] + lengths[-1]
            lengths[0] = extra // 2
            lengths[-1] = extra - lengths[0]
            if len(lengths) == 2:
                return (lengths[0], lengths[-1])
    return tuple(lengths)


def iter_slots(
    *,
    min_x: int,
    min_y: int,
    max_x: int,
    max_y: int,
    plank_length_mm: int,
    plank_width_mm: int,
    min_row_width_mm: int,
    min_end_mm: int,
    stagger: Literal["third", "half"],
    direction: Literal["along_x", "along_y"],
) -> tuple[Slot, ...]:
    if direction == "along_x":
        along = max_x - min_x
        across = max_y - min_y
    else:
        along = max_y - min_y
        across = max_x - min_x
    rows = split_rows(across, plank_width_mm, min_row_width_mm)
    step = plank_length_mm // 3 if stagger == "third" else plank_length_mm // 2
    slots: list[Slot] = []
    offset = min_y if direction == "along_x" else min_x
    for row_index, width in enumerate(rows):
        start_cut = plank_length_mm
        if row_index:
            start_cut = plank_length_mm - ((row_index * step) % plank_length_mm)
            if start_cut < min_end_mm:
                start_cut = min_end_mm
            if start_cut > plank_length_mm - min_end_mm:
                start_cut = plank_length_mm
        lengths = _row_lengths(along, plank_length_mm, start_cut, min_end_mm)
        cursor = min_x if direction == "along_x" else min_y
        for col, length in enumerate(lengths):
            if col == 0 and length != plank_length_mm:
                piece_kind: Literal["full", "start_cut", "end_cut", "rip"] = "start_cut"
            elif col == len(lengths) - 1 and length != plank_length_mm:
                piece_kind = "end_cut"
            else:
                piece_kind = "full"
            if width != plank_width_mm:
                piece_kind = "rip"
            if direction == "along_x":
                slots.append(
                    Slot(row_index, cursor, offset, cursor + length, offset + width, piece_kind)
                )
            else:
                slots.append(
                    Slot(row_index, offset, cursor, offset + width, cursor + length, piece_kind)
                )
            cursor += length
        offset += width
    return tuple(slots)

File: Kernel/Sources/PlnFlrLayout/Grid.swift
struct Slot: Equatable {
    var kind: PieceKind
    var rowIndex: Int
    var x0: Int
    var x1: Int
    var y0: Int
    var y1: Int
}

public func splitRows(innerMm: Int, plankWidthMm: Int, minRowWidthMm: Int) throws -> [Int] {
    guard innerMm > 0, plankWidthMm > 0 else { throw LayoutError.spansMustBePositive }
    let n = innerMm / plankWidthMm
    let rem = innerMm % plankWidthMm
    if n == 0 { return [innerMm] }
    if rem == 0 { return Array(repeating: plankWidthMm, count: n) }
    if rem < minRowWidthMm {
        let extra = plankWidthMm + rem
        let first = extra / 2
        let last = extra - first
        let middle = n - 1
        if middle < 0 {
            return last == 0 ? [first] : [first, last]
        }
        return [first] + Array(repeating: plankWidthMm, count: middle) + [last]
    }
    return Array(repeating: plankWidthMm, count: n) + [rem]
}

func rowLengths(innerMm: Int, plankLengthMm: Int, startCutMm: Int, minEndMm: Int) -> [Int] {
    var start = startCutMm
    if start <= 0 || start >= plankLengthMm {
        start = plankLengthMm
    }
    var remaining = innerMm - start
    if remaining < 0 { return [innerMm] }
    var lengths = [start]
    while remaining > 0 {
        if remaining >= plankLengthMm {
            lengths.append(plankLengthMm)
            remaining -= plankLengthMm
        } else {
            lengths.append(remaining)
            remaining = 0
        }
    }
    if lengths.count >= 2, lengths[lengths.count - 1] < minEndMm {
        let need = minEndMm - lengths[lengths.count - 1]
        if lengths[0] - need >= minEndMm {
            lengths[0] -= need
            lengths[lengths.count - 1] += need
        } else {
            let extra = lengths[0] + lengths[lengths.count - 1]
            lengths[0] = extra / 2
            lengths[lengths.count - 1] = extra - lengths[0]
        }
    }
    return lengths
}

func iterSlots(
    minX: Int,
    minY: Int,
    maxX: Int,
    maxY: Int,
    plankLengthMm: Int,
    plankWidthMm: Int,
    minRowWidthMm: Int,
    minEndMm: Int,
    stagger: Stagger,
    direction: Axis
) throws -> [Slot] {
    let along = direction == .alongX ? maxX - minX : maxY - minY
    let across = direction == .alongX ? maxY - minY : maxX - minX
    let rows = try splitRows(innerMm: across, plankWidthMm: plankWidthMm, minRowWidthMm: minRowWidthMm)
    let step = stagger == .third ? plankLengthMm / 3 : plankLengthMm / 2
    var slots: [Slot] = []
    var offset = direction == .alongX ? minY : minX
    for (rowIndex, width) in rows.enumerated() {
        var startCut = plankLengthMm
        if rowIndex != 0 {
            startCut = plankLengthMm - ((rowIndex * step) % plankLengthMm)
            if startCut < minEndMm { startCut = minEndMm }
            if startCut > plankLengthMm - minEndMm { startCut = plankLengthMm }
        }
        let lengths = rowLengths(
            innerMm: along,
            plankLengthMm: plankLengthMm,
            startCutMm: startCut,
            minEndMm: minEndMm
        )
        var cursor = direction == .alongX ? minX : minY
        for (col, length) in lengths.enumerated() {
            var kind: PieceKind
            if col == 0, length != plankLengthMm {
                kind = .startCut
            } else if col == lengths.count - 1, length != plankLengthMm {
                kind = .endCut
            } else {
                kind = .full
            }
            if width != plankWidthMm { kind = .rip }
            if direction == .alongX {
                slots.append(
                    Slot(
                        kind: kind,
                        rowIndex: rowIndex,
                        x0: cursor,
                        x1: cursor + length,
                        y0: offset,
                        y1: offset + width
                    )
                )
            } else {
                slots.append(
                    Slot(
                        kind: kind,
                        rowIndex: rowIndex,
                        x0: offset,
                        x1: offset + width,
                        y0: cursor,
                        y1: cursor + length
                    )
                )
            }
            cursor += length
        }
        offset += width
    }
    return slots
}
